lib/SILOptimizer/Transforms/AllocBoxToStack.cpp

```diff
@@ -1213,9 +1213,51 @@ static void rewriteApplySites(AllocBoxToStackState &pass) {
     auto *FRI = cast<FunctionRefInst>(Apply.getCallee());
     Apply.getInstruction()->eraseFromParent();
 
-    // TODO: Erase from module if there are no more uses.
-    if (FRI->use_empty())
+    if (FRI->use_empty()) {
+      auto referencedFn = FRI->getReferencedFunction();
       FRI->eraseFromParent();
+
+      // TODO: Erase from module if there are no more uses.
+      // If the function has no remaining references, it should eventually
+      // be deleted. We can't do that from a function pass, since the function
```

Comment:
@eeckstein I'm surprised that we can't mark functions deleted (which should automatically remove the function body). We should even have a `SILFunction::deleteIfUnused` API. Do you remember why we never did that?

Reply:
Well, it's not that easy to see if a function is unused. There is no use list for functions. We have `Refcount`, but that also includes references from utilities, like the cache in the Deserializer. Also, in case of cycles the refcount would be useless.
So the strategy is to keep unused functions around until the DeadFunctionElimination pass deletes them eventually.